### PR TITLE
[DEVEX-146] Creation of a GitHub Action for Terraform Drift Detection

### PR DESCRIPTION
**Pull request by mamu0**
### List of changes

Added a new GitHub action into file infra_drift_detection.yml, the action can be started manually or scheduled to run every day at 8 am.
The GitHub Action is responsible for drift detecting, via a terraform plan, doing a check  between the terraform state and the resources deployed on Azure.
If there are changes to the state, the pipe will fail, and if the Slack Webhook URL is present, it will send a notification with the information.
However, if there are no drifts the pipe will finish successfully without sending notifications.

### Motivation and context

With this template we simplify the addition of the drift check, using the new infrastructure.

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### Other information
I updated the repo by adding the dev folder in infra and the identity and repository folders following the [guide ](https://pagopa.atlassian.net/wiki/spaces/DevEx/pages/1035272271/Configurazione+di+un+repository+GitHub)

### CODE REVIEW

**Comment by gunzip** [comment]
can you change this PR title adding a reference to the jira issue here?

https://pagopa.atlassian.net/jira/software/c/projects/DEVEX/boards/501/timeline

(create one if it doesn't exists yet)

**Comment by gunzip** [comment]
why draft?

**Comment by mamu0** [comment]
> why draft?

Because I wasn't sure of the final request, @Krusty93 certainly knows it in more detail, I have now configured the "unconventional" draft detection, because instead of checking only the changes on the AZURE console side that are inconsistent with the status, I also check the changes on the terraform code side that are not applied.

Now it work, but I don't know if it needs to be changed to remove control over unapplied terraform code.

File: .github/workflows/infra_drift_detection.yml [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+
+env:
+  ARM_SUBSCRIPTION_ID: ${{ secrets.ARM_SUBSCRIPTION_ID }}
+  ARM_TENANT_ID: ${{ secrets.ARM_TENANT_ID }}
+  ARM_USE_OIDC: true
+  ARM_USE_AZUREAD: true
+  ARM_STORAGE_USE_AZUREAD: true
+  BASE_DIR: ${{ inputs.base_path != '' && inputs.base_path || 'infra/resources/' }}
+  AZURE_ENVIRONMENT: ${{ inputs.environment != '' && inputs.environment || 'dev' }}
```
Comment (Krusty93) [inline .github/workflows/infra_drift_detection.yml:40]: As we don't have a `dev` environment, I would stick to use `prod` as default. What do you think?

File: .github/workflows/infra_drift_detection.yml [hunk 2]
```diff
@@ -0,0 +1,184 @@
+name: Drift Detection
+
+on:
+  workflow_dispatch:
+    inputs:
+      environment:
+        description: Environment where the resources will be deployed.
+        type: string
+        required: true
+        default: dev
```
Comment (Krusty93) [inline .github/workflows/infra_drift_detection.yml:10]: ditto

File: .github/workflows/infra_drift_detection.yml [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+      # The plan output is saved in a file and then processed to remove unnecessary lines
+      # The step never fails but the result is checked in the next step
+      # This is because we want to post the plan output in the PR even if the plan fails
+      - name: Terraform Plan
+        id: plan
+        working-directory: ${{ steps.directory.outputs.dir }}
+        run: |
+          terraform plan -lock-timeout=3000s -no-color 2>&1 | tee plan_output.txt
+
+          OUTPUT=$(grep -Ev "Refreshing state|state lock|Reading|Read" plan_output.txt | tail -c 60000)
```
Comment (Krusty93) [inline .github/workflows/infra_drift_detection.yml:129]: The entire line could be removed as `tail -c 60000` is been introduced due to a length limit in PR comments, and for the same reason redundant strings have been rid of

File: .github/workflows/infra_drift_detection.yml [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+          
+      # Run Terraform Plan
+      # The plan output is saved in a file and then processed to remove unnecessary lines
+      # The step never fails but the result is checked in the next step
+      # This is because we want to post the plan output in the PR even if the plan fails
+      - name: Terraform Plan
+        id: plan
+        working-directory: ${{ steps.directory.outputs.dir }}
+        run: |
+          terraform plan -lock-timeout=3000s -no-color 2>&1 | tee plan_output.txt
```
Comment (Krusty93) [inline .github/workflows/infra_drift_detection.yml:127]: Should we stick to use this approach or we can leverage on the file generated by TF? Something like `terraform plan -out=tfplan.json` and eventually including the flag `-detailed-exitcode` to perform the check
What's your opinion?

File: infra/resources/dev/main.tf [hunk 5]
```diff
@@ -0,0 +1,21 @@
+terraform {
```
Comment (Krusty93) [inline infra/resources/dev/main.tf:1]: Can we remove this TF config?

File: infra/repository/locals.tf [hunk 6]
```diff
@@ -0,0 +1,22 @@
+locals {
+  project = "dx-d"
```
Comment (Krusty93) [inline infra/repository/locals.tf:2]: We only have `dx-p` or am I missing something?

**Comment by Krusty93** [comment]
> > why draft?
> 
> Because I wasn't sure of the final request, @Krusty93 certainly knows it in more detail, I have now configured the "unconventional" draft detection, because instead of checking only the changes on the AZURE console side that are inconsistent with the status, I also check the changes on the terraform code side that are not applied.
> 
> Now it work, but I don't know if it needs to be changed to remove control over unapplied terraform code.

Very nice work!
My personal opinion is to stick to this behavior, in order to help developers to eventually address code issues and/or remind/make clear them to apply code actually merged in the `main` branch.

On the other hand, the `-refresh-only` flag could help us to avoid "fake drifts" detection (i.e. cosmos db indexes)

File: .github/workflows/infra_drift_detection.yml (hunk 1 above)
Comment (mamu0) [reply .github/workflows/infra_drift_detection.yml:40]: I don't think there are any problems, I used `dev` just to use a test environment. But now I change the various references to `prod`

File: .github/workflows/infra_drift_detection.yml [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    name: Terraform Drift Detection
+    runs-on: ${{ inputs.use_private_agent == true && 'self-hosted' || 'ubuntu-20.04' }}
+    environment: ${{ inputs.environment != '' && inputs.environment || 'prod' }}-ci
+    concurrency:
+      group: ${{ github.workflow }}-ci
+      cancel-in-progress: true
+    permissions:
+      id-token: write
+      contents: read
+      pull-requests: write
```
Comment (Krusty93) [inline .github/workflows/infra_drift_detection.yml:53]: This permission can be removed

File: infra/identity/dev/westeurope/locals.tf [hunk 8]
```diff
@@ -0,0 +1,18 @@
+locals {
```
Comment (Krusty93) [inline infra/identity/dev/westeurope/locals.tf:1]: Please move files one level upper, deleting `westeurope` folder

File: .github/workflows/infra_drift_detection.yml (hunk 7 above)
Comment (mamu0) [reply .github/workflows/infra_drift_detection.yml:53]: all 3?

File: infra/resources/dev/locals.tf [hunk 9]
```diff
@@ -0,0 +1,12 @@
+locals {
+  env_short = "d"
+  location  = "westeurope"
+
+  tags = {
+    CostCenter  = "TS310 - PAGAMENTI & SERVIZI"
+    CreatedBy   = "Terraform"
+    Environment = "dev"
+    Owner       = "DevEx"
+    Source      = "https://github.com/pagopa/dx/blob/main/infra/resources/dev/westeurope"
```
Comment (Krusty93) [inline infra/resources/dev/locals.tf:10]: ditto

File: .github/workflows/infra_drift_detection.yml (hunk 7 above)
Comment (Krusty93) [reply .github/workflows/infra_drift_detection.yml:53]: no just `pull-requests`

File: infra/repository/main.tf [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    github = {
+      source  = "integrations/github"
+      version = "6.1.0"
+    }
+  }
+
+  backend "azurerm" {
+    resource_group_name  = "terraform-state-rg"
+    storage_account_name = "tfdevdx"
```
Comment (Krusty93) [inline infra/repository/main.tf:17]:
```suggestion
    storage_account_name = "tfproddx"
```

Comment (Krusty93) [reply infra/repository/main.tf:17]: Remember to import this repository into the state
`terraform import github_repository.this dx`

**Review by Krusty93 — APPROVED** [review]

**Pull request merged**